src/peer_manager.rs

```
@@ -6,17 +6,97 @@
+    pub fn our_info(&self) -> &BTreeMap<u64, Hash> {
+        &self.peers[self.our_id.public_id()]
+    }
+
+    /// Returns peer info: created events index and hash.
+    pub fn peer_info(&self, peer_id: &S::PublicId) -> Option<&BTreeMap<u64, Hash>> {
+        self.peers.get(peer_id)
+    }
+
+    /// Returns all sorted peer_ids.
+    pub fn all_ids(&self) -> Vec<&S::PublicId> {
+        self.peers.keys().collect()
+    }
+
+    /// Returns an iterator of peers.
+    pub fn iter(&self) -> btree_map::Iter<S::PublicId, BTreeMap<u64, Hash>> {
+        self.peers.iter()
+    }
+
+    /// Add a peer into the map.
+    pub fn add_peer(&mut self, peer_id: S::PublicId) {
+        let _ = self.peers.entry(peer_id).or_insert_with(BTreeMap::new);
+    }
+
+    /// Check whether the input count becomes the super majority of the network.
+    pub fn is_super_majority(&self, count: usize) -> bool {
+        3 * count > 2 * self.peers.len()
+    }
+
+    /// Returns the index of the last event created by this peer. Returns `None` if cannot find.
+    pub fn last_event_index(&self, peer_id: &S::PublicId) -> Option<u64> {
+        self.peers
+            .get(peer_id)
+            .and_then(|events| events.keys().rev().next().cloned())
+    }
+
+    /// Returns the hash of the last event created by this peer. Returns `None` if cannot find.
+    pub fn last_event_hash(&self, peer_id: &S::PublicId) -> Option<&Hash> {
+        self.peers
+            .get(peer_id)
+            .and_then(|events| events.values().rev().next())
+    }
+
+    /// Returns the hash of the indexed event.
+    pub fn event_by_index(&self, peer_id: &S::PublicId, index: u64) -> Option<&Hash> {
+        self.peers
+            .get(peer_id)
+            .and_then(|events| events.get(&index))
+    }
+
+    /// Add event created by the peer. Returns an error if `peer` is not known, `Ok(None)` if the
+    /// `index` wasn't previously held for that peer, or `Ok(Some(hash))` if it was previously held
```

Why not return an error if we get something from the peer with exactly the same index but a different event for that index (i.e. hash doesn't match) ? That seems more like a malice/error case than normal and we shouldn't be doing work for that (even if it's replacement of previous value).

---

cool, merging this then